core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,546 @@
+  }
+
+  def submitDriver(desc: MesosDriverDescription): CreateSubmissionResponse = {
+    val c = new CreateSubmissionResponse
+    val submitDate = new Date()
+    desc.submissionId = Some(newDriverId(submitDate))
+    desc.submissionDate = Some(submitDate)
+
+    stateLock.synchronized {
+      if (queue.isFull) {
+        c.success = false
+        c.message = "Already reached maximum submission size"
+      }
+      c.submissionId = desc.submissionId.get
+      queue.offer(desc)
+      c.success = true
+    }
+    c
+  }
+
+  def killDriver(submissionId: String): KillSubmissionResponse = {
+    val k = new KillSubmissionResponse
+    k.submissionId = submissionId
+    stateLock.synchronized {
+      // We look for the requested driver in the following places:
+      // 1. Check if submission is running or launched.
+      // 2. Check if it's still queued.
+      // 3. Check if it's in the retry list.
+      // 4. Check if it has already completed.
+      if (launchedDrivers.contains(submissionId)) {
+        val task = launchedDrivers.get(submissionId)
+        driver.killTask(task.taskId)
+        k.success = true
+        k.message = "Killing running driver"
+      } else if (queue.remove(submissionId)) {
+        k.success = true
+        k.message = "Removed driver while it's still pending"
+      } else if (superviseRetryList.remove(submissionId)) {
+        k.success = true
+        k.message = "Removed driver while it's retrying"
+      } else if (finishedDrivers.exists(s => s.submission.submissionId.equals(submissionId))) {
+        k.success = false
+        k.message = "Driver already terminated"
+      } else {
+        k.success = false
+        k.message = "Cannot find driver"
+      }
+    }
+    k
+  }
+
+  /**
+   * Recover scheduler state that is persisted.
+   * We still need to do task reconciliation to be up to date of the latest task states
+   * as it might have changed while the scheduler is failing over.
+   */
+  def recoverState: Unit = {
+    stateLock.synchronized {
+      queue = new DriverQueue(engineFactory.createEngine("driverQueue"), queuedCapacity)
+      launchedDrivers = new LaunchedDrivers(engineFactory.createEngine("launchedDrivers"))
+      // There is potential timing issue where a queued driver might have been launched
+      // but the scheduler shuts down before the queued driver was able to be removed
+      // from the queue. We try to mitigate this issue by walking through all queued drivers
+      // and remove if they're already launched.
+      queue.drivers.map(_.submissionId.get).filter(launchedDrivers.contains).foreach(queue.remove)
+
+      superviseRetryList = new SuperviseRetryList(engineFactory.createEngine("retryList"))
+      // TODO: Consider storing finished drivers so we can show them on the UI after
+      // failover. For now we clear the history on each recovery.
+      finishedDrivers.clear()
+    }
+  }
+
+  def start(): Unit = {
```

Comment:
There seems to be some necessary dependency order here. IIUC `registered` gets called first, then `start()`, which then calls `recoverState()` before we can call any other method like `submitDriver`, otherwise we'll get NPE. We should explicitly state this order somewhere in the comments